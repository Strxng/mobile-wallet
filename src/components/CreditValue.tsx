import React from "react";
import { Text, View } from "react-native";
import FontAwesome from "@expo/vector-icons/FontAwesome";

interface CreditValueProps {
  value: number;
}

export const CreditValue = ({ value }: CreditValueProps) => {
  const color = value === 0 ? "#52525b" : value > 0 ? "#079669" : "#9f1339";

  const iconName =
    value === 0 ? "minus" : value > 0 ? "arrow-up" : "arrow-down";

  const normalizedValue =
    value < 0 ? (value * -1).toFixed(2) : value.toFixed(2);
  const valueString = normalizedValue.toString().replace(".", ",");

  return (
    <View
      testID="credit-value"
      className="flex flex-row justify-center items-center"
    >
      <View
        testID="credit-value-icon-container"
        style={{ backgroundColor: color }}
        className="rounded-full h-5 w-5 justify-center items-center"
      >
        <FontAwesome size={12} name={iconName} color={"#FFF"} />
      </View>

      <Text
        testID="credit-value-text"
        style={{ color: color }}
        className="text-lg ml-1 font-semibold"
      >
        {`R$${valueString}`}
      </Text>
    </View>
  );
};
