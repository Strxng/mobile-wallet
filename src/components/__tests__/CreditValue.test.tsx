import React from "react";
import { CreditValue } from "../CreditValue";
import { render, cleanup } from "react-native-testing-library";

afterEach(cleanup);

describe("<CreditValue />", () => {
  it("should match snapshot", () => {
    const rendered = render(<CreditValue value={0} />).toJSON();
    expect(rendered).toMatchSnapshot();
  });

  it("should render a label with one icon and one text", () => {
    const rendered = render(<CreditValue value={0} />);
    const creditValueRendered = rendered.getByTestId("credit-value");

    expect(creditValueRendered.props.children.length).toBe(2);
  });

  it("should render a text with the same value provided", () => {
    const value = 0;

    const rendered = render(<CreditValue value={value} />);
    const textRendered = rendered.getByTestId("credit-value-text");

    expect(textRendered.children).toEqual([
      `R$${value.toFixed(2).toString().replace(".", ",")}`,
    ]);
  });

  it("should render a creaditValue label with the correct color when value is positive", () => {
    const value = 1;

    const rendered = render(<CreditValue value={value} />);
    const textRendered = rendered.getByTestId("credit-value-text");
    const iconContainerRendered = rendered.getByTestId(
      "credit-value-icon-container"
    );

    let textColor = "";
    let iconColor = "";

    textRendered.props.style.forEach((styleObj: any) => {
      Object.keys(styleObj).forEach((prop) => {
        if (prop === "color") {
          iconColor = styleObj[prop];
        }
      });
    });

    iconContainerRendered.props.style.forEach((styleObj: any) => {
      Object.keys(styleObj).forEach((prop) => {
        if (prop === "backgroundColor") {
          console.log("achou", styleObj[prop]);
          textColor = styleObj[prop];
        }
      });
    });

    expect(iconColor).toBe("#079669");
    expect(textColor).toBe("#079669");
  });

  it("should render a creaditValue label with the correct color when value is negative", () => {
    const value = -1;

    const rendered = render(<CreditValue value={value} />);
    const textRendered = rendered.getByTestId("credit-value-text");
    const iconContainerRendered = rendered.getByTestId(
      "credit-value-icon-container"
    );

    let textColor = "";
    let iconColor = "";

    textRendered.props.style.forEach((styleObj: any) => {
      Object.keys(styleObj).forEach((prop) => {
        if (prop === "color") {
          iconColor = styleObj[prop];
        }
      });
    });

    iconContainerRendered.props.style.forEach((styleObj: any) => {
      Object.keys(styleObj).forEach((prop) => {
        if (prop === "backgroundColor") {
          console.log("achou", styleObj[prop]);
          textColor = styleObj[prop];
        }
      });
    });

    expect(iconColor).toBe("#9f1339");
    expect(textColor).toBe("#9f1339");
  });

  it("should render a creaditValue label with the correct color when value is neutral", () => {
    const value = 0;

    const rendered = render(<CreditValue value={value} />);
    const textRendered = rendered.getByTestId("credit-value-text");
    const iconContainerRendered = rendered.getByTestId(
      "credit-value-icon-container"
    );

    let textColor = "";
    let iconColor = "";

    textRendered.props.style.forEach((styleObj: any) => {
      Object.keys(styleObj).forEach((prop) => {
        if (prop === "color") {
          iconColor = styleObj[prop];
        }
      });
    });

    iconContainerRendered.props.style.forEach((styleObj: any) => {
      Object.keys(styleObj).forEach((prop) => {
        if (prop === "backgroundColor") {
          console.log("achou", styleObj[prop]);
          textColor = styleObj[prop];
        }
      });
    });

    expect(iconColor).toBe("#52525b");
    expect(textColor).toBe("#52525b");
  });
});
