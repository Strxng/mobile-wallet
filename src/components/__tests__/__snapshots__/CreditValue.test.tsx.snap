// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreditValue /> should match snapshot 1`] = `
<View
  style={
    [
      {
        "display": "flex",
      },
      {
        "flexDirection": "row",
      },
      {
        "justifyContent": "center",
      },
      {
        "alignItems": "center",
      },
    ]
  }
  testID="credit-value"
>
  <View
    style={
      [
        {
          "borderBottomLeftRadius": 9999,
          "borderBottomRightRadius": 9999,
          "borderTopLeftRadius": 9999,
          "borderTopRightRadius": 9999,
        },
        {
          "height": 20,
        },
        {
          "width": 20,
        },
        {
          "justifyContent": "center",
        },
        {
          "alignItems": "center",
        },
        {
          "backgroundColor": "#52525b",
        },
      ]
    }
    testID="credit-value-icon-container"
  >
    <Text />
  </View>
  <Text
    style={
      [
        {
          "fontSize": 18,
          "lineHeight": 28,
        },
        {
          "marginLeft": 4,
        },
        {
          "fontWeight": "600",
        },
        {
          "color": "#52525b",
        },
      ]
    }
    testID="credit-value-text"
  >
    R$0,00
  </Text>
</View>
`;
